crates/bin/prove_block/src/main.rs

```diff
@@ -0,0 +1,741 @@
+    let rpc_storage = RpcStorage::new(pathfinder_client, provider_url, block_number);
+    let cached_storage = CachedRpcStorage::new(Default::default(), rpc_storage);
+
+    // TODO: nasty clone, the conversion fns don't take references
+    let transactions: Vec<_> =
+        block_with_txs.transactions.clone().into_iter().map(starknet_rs_tx_to_internal_tx).collect();
+
+    // TODO: these maps that we pass in to build_initial_state() are built only on items from the
+    // state diff, but we will need all items accessed in any way during the block (right?) which
+    // probably means filling in the missing details with API calls
+    let mut address_to_class_hash: HashMap<_, _> = state_update
+        .state_diff
+        .deployed_contracts
+        .iter()
+        .map(|contract| (contract.address, contract.class_hash))
+        .collect();
+
+    let address_to_nonce = state_update
+        .state_diff
+        .nonces
+        .iter()
+        .map(|nonce_update| {
+            // derive original nonce
+            let num_nonce_bumps =
+                Felt252::from(transactions.iter().fold(0, |acc, tx| {
+                    acc + if tx.sender_address == Some(nonce_update.contract_address) { 1 } else { 0 }
+                }));
+            assert!(nonce_update.nonce > num_nonce_bumps);
+            let previous_nonce = nonce_update.nonce - num_nonce_bumps;
+            log::debug!(
+                "probably-account contract {} nonce: {} - {} => {}",
+                nonce_update.contract_address,
+                nonce_update.nonce,
+                num_nonce_bumps,
+                previous_nonce,
+            );
+            (nonce_update.contract_address, previous_nonce)
+        })
+        .collect();
+
+    let class_hash_to_compiled_class_hash: HashMap<_, _> = state_update
+        .state_diff
+        .declared_classes
+        .iter()
+        .map(|class| (class.class_hash, class.compiled_class_hash))
+        .collect();
+
+    let storage_updates = state_update
+        .state_diff
+        .storage_diffs
+        .iter()
+        .map(|diffs| {
+            let storage_entries = diffs.storage_entries.iter().map(|e| (e.key, e.value)).collect();
+            (diffs.address, storage_entries)
+        })
+        .collect();
+
+    // TODO: avoid expensive clones here, probably by letting build_initial_state() take references
+    let (accessed_contracts, mut initial_state) = build_initial_state(
+        FactFetchingContext::new(cached_storage),
+        &provider,
+        block_number,
+        address_to_class_hash.clone(),
+        address_to_nonce,
+        class_hash_to_compiled_class_hash.clone(),
+        storage_updates,
+    )
+    .await?;
+
+    // fill in class hashes for each accessed contract
+    for address in &accessed_contracts {
+        if *address != Felt252::ONE && !address_to_class_hash.contains_key(address) {
```

Comment:
Shouldn't we do the same for address 0? Would be good to encapsulate the condition in a function to make this a bit clearer.